File: src/main/java/cz/wa/secretagent/view/model/WeaponModel.java
package cz.wa.secretagent.view.model;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;

import cz.wa.secretagent.view.SAMGraphics;
import cz.wa.secretagent.view.TileId;
import cz.wa.secretagent.view.texture.TextureToDraw;
import cz.wa.wautils.math.Rectangle2D;

/**
 * Model of weapon, consists of two models, one for wielding the weapon, second for firing 
 * 
 * @author Ondrej Milenovsky
 */
public class WeaponModel extends AbstractModel {
    private final AnimatedModel wieldModel;
    private final Vector2D wieldModelCenter;
    private final AnimatedModel fireModel;
    private final Vector2D fireModelCenter;

    private transient int hash;

    public WeaponModel(AnimatedModel wieldModel, Vector2D wieldModelCenter, AnimatedModel fireModel,
            Vector2D fireModelCenter, double scale) {
        super(scale);
        this.wieldModel = wieldModel;
        this.wieldModelCenter = wieldModelCenter;
        this.fireModel = fireModel;
        this.fireModelCenter = fireModelCenter;
    }

    @Override
    public Rectangle2D getMaxBounds() {
        return MAX_BOUNDS;
    }

    public AnimatedModel getModel(boolean firing) {
        if (firing) {
            return fireModel;
        } else {
            return wieldModel;
        }
    }

    public Vector2D getCenter(boolean firing) {
        if (firing) {
            return fireModelCenter;
        } else {
            return wieldModelCenter;
        }
    }

    @Override
    public ModelType getType() {
        return ModelType.WEAPON;
    }

    @Override
    public Set<TileId> getAllTileIds() {
        Set<TileId> ret = new HashSet<TileId>();
        ret.addAll(wieldModel.getAllTileIds());
        ret.addAll(fireModel.getAllTileIds());
        return ret;
    }

    @Override
    public boolean hasLinkedTextures() {
        return wieldModel.hasLinkedTextures() && fireModel.hasLinkedTextures();
    }

    @Override
    protected Rectangle2D linkTexturesInternal(SAMGraphics graphics) {
        wieldModel.linkTextures(graphics);
        fireModel.linkTextures(graphics);
        return wieldModel.getBounds();
    }

    @Override
    protected Collection<TextureToDraw> getAllTextures() {
        List<TextureToDraw> ret = new ArrayList<TextureToDraw>();
        ret.addAll(wieldModel.getAllTextures());
        ret.addAll(fireModel.getAllTextures());
        return ret;
    }

    @Override
    public int hashCode() {
        if (hash == 0) {
            final int prime = 31;
            int result = super.hashCode();
            result = prime * result + ((fireModel == null) ? 0 : fireModel.hashCode());
            result = prime * result + ((fireModelCenter == null) ? 0 : fireModelCenter.hashCode());
            result = prime * result + ((wieldModel == null) ? 0 : wieldModel.hashCode());
            result = prime * result + ((wieldModelCenter == null) ? 0 : wieldModelCenter.hashCode());
            hash = result;
        }
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (!super.equals(obj))
            return false;
        if (getClass() != obj.getClass())
            return false;
        WeaponModel other = (WeaponModel) obj;
        if (fireModel == null) {
            if (other.fireModel != null)
                return false;
        } else if (!fireModel.equals(other.fireModel))
            return false;
        if (fireModelCenter == null) {
            if (other.fireModelCenter != null)
                return false;
        } else if (!fireModelCenter.equals(other.fireModelCenter))
            return false;
        if (wieldModel == null) {
            if (other.wieldModel != null)
                return false;
        } else if (!wieldModel.equals(other.wieldModel))
            return false;
        if (wieldModelCenter == null) {
            if (other.wieldModelCenter != null)
                return false;
        } else if (!wieldModelCenter.equals(other.wieldModelCenter))
            return false;
        return true;
    }

}

File: src/main/kotlin/secretAgent/view/model/WeaponModel.kt
package secretAgent.view.model

import cz.wa.secretagent.view.model.AbstractModel
import cz.wa.wautils.math.Rectangle2D
import org.apache.commons.math3.geometry.euclidean.twod.Vector2D
import secretAgent.view.SamGraphics
import secretAgent.view.renderer.TextureToDraw
import secretAgent.view.renderer.TileId
import secretAgent.world.ModelType
import java.util.ArrayList
import java.util.HashSet

/**
 * Model of weapon, consists of two models, one for wielding the weapon, second for firing
 *
 * @author Ondrej Milenovsky
 */
class WeaponModel(
        private val wieldModel: AnimatedModel,
        private val wieldModelCenter: Vector2D,
        private val fireModel: AnimatedModel,
        private val fireModelCenter: Vector2D, scale: Double) : AbstractModel(scale) {

    @Transient
    private var hash = 0

    override val maxBounds: Rectangle2D
        get() = MAX_BOUNDS

    override val type: ModelType
        get() = ModelType.WEAPON

    override val allTileIds: Set<TileId>
        get() {
            val ret = HashSet<TileId>()
            ret += wieldModel.allTileIds
            ret += fireModel.allTileIds
            return ret
        }

    override fun getAllTextures(): Collection<TextureToDraw> {
        val ret = ArrayList<TextureToDraw>()
        ret += wieldModel.allTextures!!
        ret += fireModel.allTextures!!
        return ret
    }

    fun getModel(firing: Boolean): AnimatedModel = if (firing) fireModel else wieldModel

    fun getCenter(firing: Boolean): Vector2D = if (firing) fireModelCenter else wieldModelCenter

    override fun hasLinkedTextures(): Boolean = wieldModel.hasLinkedTextures() && fireModel.hasLinkedTextures()

    override fun linkTexturesInternal(graphics: SamGraphics): Rectangle2D? {
        wieldModel.linkTextures(graphics)
        fireModel.linkTextures(graphics)
        return wieldModel.bounds
    }

    override fun hashCode(): Int {
        if (hash == 0) {
            val prime = 31
            var result = super.hashCode()
            result = prime * result + fireModel.hashCode()
            result = prime * result + fireModelCenter.hashCode()
            result = prime * result + wieldModel.hashCode()
            result = prime * result + wieldModelCenter.hashCode()
            hash = result
        }
        return hash
    }

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true
        if (!super.equals(other))
            return false
        return other is WeaponModel && fireModel == other.fireModel && fireModelCenter == other.fireModelCenter &&
                wieldModel == other.wieldModel && wieldModelCenter == other.wieldModelCenter
    }
}